tvb/recon/flow/surfaces_to_structural_datasets.py

```diff
@@ -20,6 +21,131 @@ class Hemisphere(enum.Enum):
+    return vertex_triangles
+
+
+def compute_triangle_normals(triangles, vertices):
+    """Calculates triangle normals."""
+    tri_u = vertices[triangles[:, 1], :] - vertices[triangles[:, 0], :]
+    tri_v = vertices[triangles[:, 2], :] - vertices[triangles[:, 0], :]
+    tri_norm = np.cross(tri_u, tri_v)
+
+    try:
+        triangle_normals = tri_norm / np.sqrt(np.sum(tri_norm ** 2, axis=1))[:, np.newaxis]
+    except FloatingPointError:
+        # TODO: NaN generation would stop execution, however for normals this case could maybe be
+        #  handled in a better way.
+        triangle_normals = tri_norm
+    return triangle_normals
+
+
+def compute_triangle_angles(vertices, number_of_triangles, triangles):
+    """
+    Calculates the inner angles of all the triangles which make up a surface
+    """
+    verts = vertices
+    # TODO: Should be possible with arrays, ie not nested loops...
+    # A short profile indicates this function takes 95% of the time to compute normals
+    # (this was a direct translation of some old matlab code)
+    angles = np.zeros((number_of_triangles, 3))
+    for tt in range(number_of_triangles):
+        triangle = triangles[tt, :]
+        for ta in range(3):
+            ang = np.roll(triangle, -ta)
+            angles[tt, ta] = np.arccos(np.dot(
+                (verts[ang[1], :] - verts[ang[0], :]) /
+                np.sqrt(np.sum((verts[ang[1], :] - verts[ang[0], :]) ** 2, axis=0)),
+                (verts[ang[2], :] - verts[ang[0], :]) /
+                np.sqrt(np.sum((verts[ang[2], :] - verts[ang[0], :]) ** 2, axis=0))))
+
+    return angles
+
+
+def compute_vertex_normals(number_of_vertices, vertex_triangles, triangles,
+                           triangle_angles, triangle_normals, vertices):
+    """
+    Estimates vertex normals, based on triangle normals weighted by the
+    angle they subtend at each vertex...
+    """
+    vert_norms = np.zeros((number_of_vertices, 3))
+    bad_normal_count = 0
+    for k in range(number_of_vertices):
+        try:
+            tri_list = list(vertex_triangles[k])
+            angle_mask = triangles[tri_list, :] == k
+            angles = triangle_angles[tri_list, :]
+            angles = angles[angle_mask][:, np.newaxis]
+            angle_scaling = angles / np.sum(angles, axis=0)
+            vert_norms[k, :] = np.mean(angle_scaling * triangle_normals[tri_list, :], axis=0)
+            # Scale by angle subtended.
+            vert_norms[k, :] = vert_norms[k, :] / np.sqrt(np.sum(vert_norms[k, :] ** 2, axis=0))
+            # Normalise to unit vectors.
+        except (ValueError, FloatingPointError):
+            # If normals are bad, default to position vector
+            # A nicer solution would be to detect degenerate triangles and ignore their
+            # contribution to the vertex normal
+            vert_norms[k, :] = vertices[k] / np.sqrt(vertices[k].dot(vertices[k]))
+            bad_normal_count += 1
```

Comment:
No, but I haven't even run it yet. The computations are mostly Tim's code that I've left as it was without changes